csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
+      ->allreduce(output_tensors, {.reduceOp = params_.redOp});
 }
 
 ReduceScatter::ReduceScatter(CommParams params)
-    : Communication(params, "reduce_scatter", false) {
-  assertBufferCount(params_.src_bufs, params_.team.size());
-  assertBufferCount(params_.dst_bufs, 1);
-}
+    : Communication(params, "reduce_scatter", false) {}
 
 c10::intrusive_ptr<c10d::Work> ReduceScatter::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list = {std::move(params_.src_bufs)};
-  auto work = comm.getBackendForTeam(params_.team, backend)
-                  ->reduce_scatter(
-                      params_.dst_bufs, buf_list, {.reduceOp = params_.redOp});
-  params_.src_bufs = std::move(buf_list.back());
-  return work;
+
+  std::vector<std::vector<at::Tensor>> input_tensors(1);
+  // FIXME: use split
+  NVF_ERROR(
+      params_.scattered_axis >= 0,
+      "scattered_axis is expected to be non-negative: ",
+      params_.scattered_axis)
+  const auto scattered_dim_size = input_tensor.size(params_.scattered_axis);
+  const auto team_size = static_cast<int64_t>(params_.team.size());
+  NVF_ERROR(
+      scattered_dim_size == team_size, scattered_dim_size, " vs ", team_size);
+  for (auto i : c10::irange(team_size)) {
+    input_tensors[0].push_back(
+        input_tensor.slice(params_.scattered_axis, i, i + 1));
+  }
+
+  std::vector<at::Tensor> output_tensors({output_tensor});
+  return comm.getBackendForTeam(params_.team, backend)
+      ->reduce_scatter(
+          output_tensors, input_tensors, {.reduceOp = params_.redOp});
 }
 
 SendRecv::SendRecv(CommParams params) : Communication(params, "send/recv") {
-  assertBuffersHaveSameSize(params_.src_bufs, params_.dst_bufs);
   NVF_ERROR(
       params_.team.size() == 1 || params_.team.size() == 2,
       "the team size should be 1 or 2");
 }
 
 c10::intrusive_ptr<c10d::Work> SendRecv::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
 
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, 1);
-    if (params_.team.size() == 1) {
-      assertBufferCount(params_.dst_bufs, 1);
-      doLocalCopy(params_.dst_bufs.at(0), params_.src_bufs.at(0));
-      return nullptr;
-    } else {
-      assertBufferCount(params_.dst_bufs, 0);
-    }
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
-    assertBufferCount(params_.dst_bufs, 1);
+  if (params_.team.size() == 1) {
+    doLocalCopy(output_tensor, input_tensor);
+    return nullptr;
   }
 
+  std::vector<at::Tensor> tensors(
+      {comm.deviceId() == params_.root ? input_tensor : output_tensor});
   return comm.sendRecv(
```

Comment:
the method `sendRecv` should eventually be removed from the class `Communicator`, as done in the other PR https://github.com/NVIDIA/Fuser/pull/2185